proposals/client-interceptors.md

```diff
@@ -0,0 +1,312 @@
+
+## Proposal
+
+### A simple example
+
+This example logs all outbound requests:
+
+```javascript
+var interceptor = function(options, nextCall) {
+  return new InterceptingCall(nextCall(options), {
+    send: function(request, next) {
+      console.log(request);
+      next(request);
+    }
+  });
+}
+
+
+var client = new MyServiceClient(hostname, credentials, { interceptors: [interceptor] });
+// ... or via method call
+client.myCall(message, {}, { interceptors: [interceptor] });
+```
+
+### Interceptor API
+
+An interceptor is a function which takes an `options` object and a `nextCall` function and returns an `InterceptingCall` object:
+
+```javascript
+/**
+ * @param {Object} options
+ * @param {string} options.hostname The server host to call
+ * @param {Object} options.credentials Credentials used to authenticate with the server
+ * @param {MethodDescriptor} options.methodDiscriptor A container of properties describing the call (see below)
+ * @param {Function} nexCall Constructs the next interceptor in the chain
+ * @returns {InterceptingCall} 
+ */
+var interceptor = function(options, nextCall) {
+  return new InterceptingCall(nexCall(options));
+}
+```
+
+An interceptor function allows developers to modify the call options, store any call-scoped values needed, and define interception methods.
+
+The interceptor function must return an `InterceptingCall` object. An `InterceptingCall` represents an element in the interceptor chain. Any intercepted methods are implemented in the requester object, an optional parameter to the constructor. Returning `new InterceptingCall(nextCall(options))` will satisfy the contract (but provide no interceptor functionality).
+
+Most interceptors will define a `requester` object (described below) which is passed to the the `InterceptingCall` constructor: `return new InterceptingCall(nextCall(options), requester);`
+
+The `options` argument to the `nextCall` function includes options accepted by `rpcCall` and `serverStreaming` constructor. Modifying the options that are passed to the `nextCall` function will have the effect of changing the options passed to the underlying `rpcCall`/`serverStreaming` constructor.
+
+Additionally, the `options` argument includes a `MethodDescriptor` object. The `MethodDescriptor` is a container for properties of the call which are used internally and may also be useful to interceptors:
+
+```javascript
+/**
+ * @param {string} name The name of the call, i.e. 'myCall' 
+ * @param {string} serviceName The name of the service, i.e. 'MyService'
+ * @param {string} path The full path of the call, i.e. '/MyService/MyCall'
+ * @param {MethodType} methodType One of two types: MethodType.UNARY or MethodType.SERVER_STREAMING
+ * @param {grpc.web.AbstractClientBase.MethodInfo} methodInfo Properties containing the serialize and deserialize functions
+ */
+MethodDescriptor(name, serviceName, path, methodType, methodInfo)
```

Comment:
wdyt about moving the existing class out of AbstractClientBase and creating grpc.web.MethodInfo that works for this and other use cases? 

Couple of reasons to consider this:
1. If grpc.web.MethodInfo also contains the path it should simplify the proto compiler as well (js_grpc_web_library).
2. It'll be confusing to have both MethodInfo and MethodDescriptor

Reply:
Yup, I think this would be a good idea. 
My initial thoughts on the implementation was to make minimal changes to the current API and CodeGen and provide backward compatibility (hence keeping `methodInfo` as-is). The reality may be that the refactor may not be compatible... we would need to make a call on backward compatibility support; if no then this would push this project to v2.0.0

Reply:
fyi, https://github.com/grpc/grpc-web/pull/567 introduced grpc.web.MethodDescriptor and MethodType.